### ISSUE TITLE
Conflict between needless_borrow and explicit_auto_deref

### ISSUE THREAD

**Issue by fintelia** [rust-lang/rust-clippy#11366]
### Summary

With code that triggers both `clippy::needless_borrow` and `clippy::explicit_auto_deref`, it seems that `clippy --fix` attempts to apply both fixes and produces invalid code.

### Reproducer

I tried this code:

```rust
pub enum DecodingBuffer<'a> {
    U8(&'a mut [u8]),
}

impl<'a> DecodingBuffer<'a> {
    pub fn as_bytes_mut(&mut self) -> &mut [u8] {
        match self {
            DecodingBuffer::U8(buf) => &mut *buf,
        }
    }
}
```

And this happened:
```
$ cargo clippy --fix
    Checking clippy-bug v0.1.0 (/home/jonathan/tmp/clippy-bug)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_bug`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0596]: cannot borrow `buf` as mutable, as it is not declared as mutable
 --> src/main.rs:8:40
  |
8 |             DecodingBuffer::U8(buf) => &mut buf,
  |                                        ^^^^^^^^
  |                                        |
  |                                        cannot borrow as mutable
  |                                        help: try removing `&mut` here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
Original diagnostics will follow.

warning: this expression creates a reference which is immediately dereferenced by the compiler
 --> src/main.rs:8:40
  |
8 |             DecodingBuffer::U8(buf) => &mut *buf,
  |                                        ^^^^^^^^^ help: change this to: `*buf`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
  = note: `#[warn(clippy::needless_borrow)]` on by default

warning: deref which would be done by auto-deref
 --> src/main.rs:8:45
  |
8 |             DecodingBuffer::U8(buf) => &mut *buf,
  |                                             ^^^^ help: try: `buf`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_auto_deref
  = note: `#[warn(clippy::explicit_auto_deref)]` on by default

warning: `clippy-bug` (bin "clippy-bug") generated 2 warnings (run `cargo clippy --fix --bin "clippy-bug"` to apply 2 suggestions)
warning: `clippy-bug` (bin "clippy-bug" test) generated 2 warnings (2 duplicates)
    Finished dev [unoptimized + debuginfo] target(s) in 0.11s
```

### Version

```text
rustc 1.73.0-nightly (e4c144684 2023-08-04)
binary: rustc
commit-hash: e4c14468460b22ac7a289a422fb7fee606e12a0b
commit-date: 2023-08-04
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 16.0.5
```
